app/src/main/java/org/stepic/droid/ui/fragments/CoursesCarouselFragment.kt

```diff
@@ -299,4 +308,48 @@ class CoursesCarouselFragment
     private fun viewAll() {
         screenManager.showCoursesList(activity, info)
     }
+
+    override fun onSuccessJoin(joinedCourse: Course) {
+        val courseIndex = courses.indexOfFirst { it.courseId == joinedCourse.courseId }
+
+        if (courseIndex >= 0) {
+            courses[courseIndex].enrollment = joinedCourse.enrollment
+            coursesRecycler.adapter.notifyItemChanged(courseIndex)
+        } else if (info.table == Table.enrolled) {
+            //insert at 0 index is more complex than just add, but order will be right
+            if (courses.isEmpty()) {
+                showCourses(mutableListOf(joinedCourse))
+            } else {
+                courses.add(0, joinedCourse)
+                coursesRecycler.adapter.notifyDataSetChanged()
+                updateOnCourseCountChanged()
+            }
+        }
+    }
+
+    private fun updateOnCourseCountChanged() {
+        fun updateSpanCount() {
```

Comment:
According to bytecode, this inner functions will be recreated every time `updateOnCourseCountChanged` was called.

```
L0
    LINENUMBER 331 L0
    NEW org/stepic/droid/ui/fragments/CoursesCarouselFragment$updateOnCourseCountChanged$1
    DUP
    ALOAD 0
    INVOKESPECIAL org/stepic/droid/ui/fragments/CoursesCarouselFragment$updateOnCourseCountChanged$1.<init> (Lorg/stepic/droid/ui/fragments/CoursesCarouselFragment;)V
    ASTORE 1
```
or
```java
Function0 updateSpanCount$ = new Function0(0) {
         // $FF: synthetic method
         // $FF: bridge method
         public Object invoke() {
            this.invoke();
            return Unit.INSTANCE;
         }

         public final void invoke() {
            if(!CoursesCarouselFragment.this.courses.isEmpty()) {
               int spanCount = Math.min(CoursesCarouselFragment.this.courses.size(), 2);
               GridLayoutManager var10000 = CoursesCarouselFragment.this.gridLayoutManager;
               if(var10000 != null) {
                  var10000.setSpanCount(spanCount);
               }

            }
         }
      };
```